libdns.go

```diff
@@ -1,394 +1,212 @@
+// [libdns.RecordAppender.AppendRecords] is called simultaneously, and two API
+// requests are made to the provider at the same time, the result of both requests
+// must be visible after they both complete; if the provider does not synchronize
+// the writing of the zone file and one request overwrites the other, then the
+// client implementation must take care to synchronize on behalf of the incompetent
+// provider. This synchronization need not be global; for example: the scope of
+// synchronization might only need to be within the same zone, allowing multiple
+// requests at once as long as all of them are for different zone. (Exact logic
+// depends on the provider.)
+//
+// Some service providers APIs may enforce rate limits or have sporadic errors.
+// It is generally expected that libdns provider packages implement basic retry
+// logic (e.g. retry up to 3-5 times with backoff in the event of a connection error
+// or some HTTP error that may be recoverable, including 5xx or 429s) when it is
+// safe to do so. Retrying/recovering from errors should not add substantial latency,
+// though. If it will take longer than a couple seconds, best to return an error.
+//
+// [Resource Record]: https://en.wikipedia.org/wiki/Domain_Name_System#Resource_records
 package libdns
 
 import (
 	"context"
-	"fmt"
-	"strconv"
 	"strings"
-	"time"
 )
 
 // [RecordGetter] can get records from a DNS zone.
-//
-// “GetRecords” returns all the records in the DNS zone.
-//
-// DNSSEC-related records are typically not included in the output, but this
-// behavior is implementation-defined. *If* an implementation includes DNSSEC
-// records in the output, this behavior should be documented.
-//
-// Implementations must honor context cancellation and be safe for concurrent
-// use.
 type RecordGetter interface {
+	// GetRecords returns all the records in the DNS zone.
+	//
+	// DNSSEC-related records are typically not included in the output, but this
+	// behavior is implementation-defined. If an implementation includes DNSSEC
+	// records in the output, this behavior should be documented.
+	//
+	// Implementations must honor context cancellation and be safe for concurrent
+	// use.
 	GetRecords(ctx context.Context, zone string) ([]Record, error)
 }
 
 // [RecordAppender] can non-destructively add new records to a DNS zone.
-//
-// “AppendRecords” creates the requested records in the given zone and returns
-// the populated records that were created. It never changes existing records.
-// Therefore, it is invalid to use this method with “CNAME”-type records.
-//
-// Implementations must honor context cancellation and be safe for concurrent
-// use.
 type RecordAppender interface {
+	// AppendRecords creates the inputted records in the given zone and returns
+	// the populated records that were created. It never changes existing records.
+	// Therefore, it is invalid to use this method with CNAME-type records.
```

Comment:
I know this was added in a previous PR, but with a revamp and same people involved in this PR: Is the intend to say that you can't add a CNAME record when you already have other types of records with the same name? Provider implementers could interpret this line to say they should just reject any attempt to add a CNAME record with AppendRecords. I would think we can just "append" a single CNAME record for a name that doesn't exist yet.

Reply:
I think the intent is that provider packages don't have to worry about this. Basically, using AppendRecords() is error-prone when appending CNAME records.

Although, as you state below, using SetRecords with CNAME is also error-prone, since another record type could already exist for the domain name.

What if we just revise the comment here to advise caution when appending CNAMEs? (Ultimately, the DNS service API should return an error, I would think, but we don't know that for sure. So it might be best if the user checks to ensure there isn't another record for the domain name already, before calling Append or Set for a CNAME record.)

Reply:
> I would think we can just "append" a single CNAME record for a name that doesn't exist yet.

Err, yes. I'm not sure how I missed that rather obvious case. 

> Is the intend to say that you can't add a CNAME record when you already have other types of records with the same name?  

Since it's not always obvious when to use `AppendRecords` and when to use `SetRecords`, my intent was to try and provide some concrete guidelines. 

> What if we just revise the comment here to advise caution when appending CNAMEs?

Sure, I'll add that in the next commit.